<!DOCTYPE HTML>
<html>
<head>
	<meta charset="utf-8" />
	<title>N.E.A.T projects</title>
	<link href="https://fonts.googleapis.com/css2?family=Varela+Round&display=swap" rel="stylesheet">
	<link rel="stylesheet" href="styles.css">
</head>
<body>
	<div class="container">
		<div class="nav-wrapper">
			<div class="left-side">
				<div class="nav-link-wrapper">
					<a href="index.html"> Home </a>
				</div>
				<div class="nav-link-wrapper">
					<a href="about.html"> About me </a>
				</div>
			</div>
			<div class="right-side">
				<div class="brand">
					<b>SRIRAM MURALIDHARAN</b>
				</div>
			</div>
		</div>
	</div>

	<div class="content-wrapper-pro">
		<h1>Automating Games Using NEAT-Algorithm</h1>

		<h2> N.E.A.T Algorithm </h2>
		<p>
			NeuroEvolution of Augmenting Topologies (NEAT) is a genetic algorithm (GA) for the generation of evolving artificial neural networks (a neuroevolution technique) developed by Ken Stanley in 2002 while at The University of Texas at Austin. It alters both the weighting parameters and structures of networks, attempting to find a balance between the fitness of evolved solutions and their diversity. It is based on applying three key techniques: tracking genes with history markers to allow crossover among topologies, applying speciation (the evolution of species) to preserve innovations, and developing topologies incrementally from simple initial structures ("complexifying").
		</p>

		<h2>Encoding</h2>
		<p>
			In biology, we have a genotype and a phenotype. A genotype is the genetic representation of a creature and the phenotype is the actualized physical representation of the creature. Evolutionary algorithms always heavily mirror biology, neuroevolution being no different in this respect.
			The question of encoding comes from the question of how do we wish to represent individuals genetically in our algorithm. The way in which we encode our individuals lays out the path for how our algorithm will handle the key evolutionary processes: selection, mutation, and crossover (also known as recombination). Any encoding will fall into one of two categories, direct or indirect.
			A direct encoding will explicitly specify everything about an individual. If it represents a neural network this means that each gene will directly be linked to some node, connection, or property of the network. This can be a binary encoding of 1s and 0s, a graph encoding (linking various nodes by weighted connections), or something even more complex. The point is that there will always be a direct connection between genotype and phenotype that is very obvious and readable.
			An indirect encoding is the exact opposite. Instead of directly specifying what a structure may look like, indirect encodings tends to specify rules or parameters of processes for creating an individual. As a result, indirect encodings are much more compact. The flip side is that setting the rules for an indirect encoding can result in a heavy bias within the search space, therefore, it is much harder to create an indirect encoding without substantial knowledge about how the encoding will be used.
			The NEAT algorithm chooses a direct encoding methodology because of this. Their representation is a little more complex than a simple graph or binary encoding, however, it is still straightforward to understand. It simply has two lists of genes, a series of nodes and a series of connections. To see what this looks like visually, I have a picture from the original paper here:
		</p>
		<center><img src="images/neat1.png" height="350" width="700"></center>
		<p>
			Input and output nodes are not evolved in the node gene list. Hidden nodes can be added or removed. As for connection nodes, they specify where a connection comes into and out of, the weight of such connection, 	whether or not the connection is enabled, and an innovation number
		</p>
		<h2>Mutation</h2>
		<p>
			In NEAT, mutation can either mutate existing connections or can add new structure to a network. If a new connection is added between a start and end node, it is randomly assigned a weight.
			If a new node is added, it is placed between two nodes that are already connected. The previous connection is disabled (though still present in the genome). The previous start node is linked to the new node with the weight of the old connection and the new node is linked to the previous end node with a weight of 1. This was found to help mitigate issues with new structural additions.
		</p>
		<center><img src="images/neat2.jpg" width="600" height="400"><img src="images/neat3.png"width="400" height="400"></center>
		<h2>Competing Conventions</h2>
		<p>
			Another big issue with evolving the topologies of neural networks is something that the NEAT paper calls “competing conventions.” The idea is that just blindly crossing over the genomes of two neural networks could result in networks that are horribly mutated and non-functional. If two networks are dependent on central nodes that both get recombined out of the network, we have an issue.

			More than that, genomes can be of different sizes. How do we align genomes that don’t seem to be obviously compatible? In biology, this is taken care of through an idea called homology. Homology is the alignment of chromosomes based on matching genes for a specific trait. Once that happens, crossover can happen with much less chance of error than if chromosomes were blindly mixed together.
			NEAT tackles this issue through the usage of historical markings (as seen above). By marking new evolutions with a historical number, when it comes time to crossover two individuals, this can be done with much less chance of creating individuals that are non-functional. Each gene can be aligned and (potentially) crossed-over. Each time a new node or new type of connection occurs, a historical marking is assigned, allowing easy alignment when it comes to breed two of our individuals. View this here:
		</p>
		<center><img src="images/neat4.png"></center>
		<h2>Speciation</h2>
		<p>
			A very interesting idea put forth in NEAT was that most new evolutions are not good ones. In fact, adding a new connection or node before any optimization of weights have occurred often leads to a lower performing individual. This puts new structures at a disadvantage. How can we protect new structures and allow them to optimize before we eliminate them from the population entirely? NEAT suggests speciation.
			Speciation simply splits up the population into several species based on the similarity of topology and connections. If the competing convention problem still existed, this would be very hard to measure! However, since NEAT uses historical markings in its encoding, this becomes much easier to measure. A function for deciding how to speciate is given in the paper, but the important part to note is that individuals in a population only have to compete with other individuals within that species. This allows for new structure to be created and optimized without fear that it will be eliminated before it can be truly explored.
			More than that, NEAT takes things one step forward through something called explicit fitness sharing. That means that individuals share how well they are doing across the species, boosting up higher performing species, though still allowing for other species to explore their structure optimization before being out evolved.
		</p>
		<h1> IMPLEMENTATIONS </h1>
		<h2> FLAP_AI: smartest flappy out there </h2>
		For source code: <a href="https://github.com/msris108/FLAP_AI"> Check out my GitHub repo </a> 
		<center>
			<video width="1000" height="600" controls>
				<source src="images/flap_AI.mp4" type="video/mp4">
				Your browser does not support the video tag.
			</video>
		</center>
		<h2> Lunar_lander_AI: A shabby simulation using pygame :)</h2>
		For source code: <a href="https://github.com/msris108/LUNAR_LANDER_AI"> Check out my GitHub repo </a> 
		<center>
			<video width="1000" height="590" controls>
				<source src="images/lunar_lander_AI.mp4" type="video/mp4">
				Your browser does not support the video tag.
			</video>
		</center>

	</div>
</body>
</html>
